### PR TITLE
Refactoring : make argument required and less Context :)

### PR DESCRIPTION
**Pull request by vdemeester**
This PR is the start of a refactoring, that does the following 🐰 :

- Clean `Project.Context` from attributes that should not be there…
- … and thus making the `Project` *API* more argument driven than context driven ; it makes it easier to use it and know what is needed for each method. It also decouple a bit the different layer from the Context.
- Move `docker/builder.go` into a separate package that can work by itself and add test to it (~90% coverage)…
- … add a `test` package with a no-op implementation of the API client, used to *fake* client calls for unit tests (should help to write more unit tests).

There is still lots of room for improvements:
- I really wonder if the variadic signatures (`services ...string`) is useful ; it made me add the options and argument before the service(s) and.. I don't really like that.
- The `docker` package can be splitted in subpackage way more (for project, service & such) so that we have less and less coupling.

But this one is pretty big already so, let's say it's phase one :smile_cat: 

/cc @joshwget @dnephin @ibuildthecloud 

🐸

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
I think there's a bit too much in this one PR, but I very much like the idea of moving away from so much crap in the Context.

File: project/types/types.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package types
```
Comment (ibuildthecloud) [inline project/types/types.go]: What is the package was options.  The everything would be `options.Up`, `options.Build`.  Seems prettier to me.

**Comment by ibuildthecloud** [comment]
I very much like this.  Also realized I should be signing my commits :)

File: project/types/types.go (hunk 1 above)
Comment (vdemeester) [inline project/types/types.go]: Completely true :-)

**Comment by vdemeester** [comment]
@ibuildthecloud updated with the package.

**Comment by vdemeester** [comment]
@ibuildthecloud I'll take your "I very much like this" as a LGTM, and we'll iterate over it anyway :wink:

**Pull request merged**